src/Extsload.sol

```diff
@@ -25,4 +25,28 @@ abstract contract Extsload is IExtsload {
 
         return value;
     }
+
+    function extsload(bytes32[] calldata slots) external view returns (bytes32[] memory) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            let free := mload(64)
+            // The offset of our array in the returndata is 32.
+            mstore(free, 32)
+            mstore(add(free, 32), slots.length)
+            // Return values will start at `free + 64`. We compute the difference between the memory
+            // and calldata offset so we don't have to track and update a separate memory offset.
+            let relativeOffset := sub(add(free, 64), slots.offset)
```

Comment:
how do you know `free + 64` is definitely bigger than slots.offset

Reply:
`slots.offset` will always be 68, 4 bytes for the function signature, 1 word for the offset, 1 word for the length

Reply:
Since this function is external and everything is in assembly, the Solidity memory convention can be discarded, meaning we can use memory starting from 0 and override `0x40`. An implementation was proposed in the following comment.
https://github.com/Uniswap/v4-core/pull/559#pullrequestreview-1982030657

Reply:
I appreciate the recognition. But I didn't mean to override your impl with mine. The comments you wrote are very useful and the way you used the relative offset instead of incrementing two variables should be more efficient. I was suggesting changing the `free` variable to a hardcode `0`. Thank you.

Reply:
@shuhuiluo i did a gas comparison of those 2 algorithms and this one is cheaper. For loading 10 sparse slots (the test in `Extsload.t.sol`), the implementation is 2023, and the old one but using `0` instead of `fmp` is 2145 :)

Reply:
@hensha256 I took the best of both worlds. Could you try this? I got 1958.
```solidity
        /// @solidity memory-safe-assembly
        assembly {
            // The abi offset of dynamic array in the returndata is 32.
            mstore(0, 0x20)
            mstore(0x20, slots.length)
            // A left bit-shift of 5 is equivalent to multiplying by 32 but costs less gas.
            let end := add(0x40, shl(5, slots.length))
            // Return values will start at 64 while calldata offset is 68.
            for { let memptr := 0x40 } 1 {} {
                // Compute calldata offset using the memory offset and store loaded value.
                mstore(memptr, sload(calldataload(add(memptr, 0x04))))
                memptr := add(memptr, 0x20)
                if iszero(lt(memptr, end)) { break }
            }
            // The end offset is also the length of the returndata.
            return(0, end)
        }
```